### PR TITLE
Consistently use TypeMismatch in TreeChecker

### PR DESCRIPTION
**Pull request by dwijnand**

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/transform/TreeChecker.scala [hunk 1]
```diff
@@ -824,16 +813,20 @@ object TreeChecker {
         && !isPrimaryConstructorReturn
         && !pt.isInstanceOf[FunOrPolyProto]
       then
-        assert(tree.tpe <:< pt, {
```
Comment (hamzaremmal) [inline compiler/src/dotty/tools/dotc/transform/TreeChecker.scala:827]: The proposed changes will also add the following check here:  `tree.tpe eq pt`. Is that ok to add ?

File: compiler/src/dotty/tools/dotc/transform/TreeChecker.scala [hunk 2]
```diff
@@ -824,16 +813,20 @@ object TreeChecker {
...
+        checkType(tree.tpe, pt, tree, "adapt")
       tree
     }
 
     override def simplify(tree: Tree, pt: Type, locked: TypeVars)(using Context): tree.type = tree
+
+    private def checkType(tp1: Type, tp2: Type, tree: untpd.Tree, step: String)(using Context) =
+      // Accept NoType <:< NoType as true
+      assert((tp1 eq tp2) || (tp1 <:< tp2), {
+        val mismatch = TypeMismatch(tp1, tp2, None)
```
Comment (hamzaremmal) [inline compiler/src/dotty/tools/dotc/transform/TreeChecker.scala:825]: Shouldn't it be ?
```suggestion
        val mismatch = TypeMismatch(tp1, tp2, Some(tree))
```

File: compiler/src/dotty/tools/dotc/transform/TreeChecker.scala (hunk 1 above)
Comment (dwijnand) [reply compiler/src/dotty/tools/dotc/transform/TreeChecker.scala:827]: Yep

File: compiler/src/dotty/tools/dotc/transform/TreeChecker.scala (hunk 2 above)
Comment (dwijnand) [reply compiler/src/dotty/tools/dotc/transform/TreeChecker.scala:825]: No, because I'm adding .className to the tree ouput.

**Pull request merged**